src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -1,100 +1,209 @@
+
+	/**
+	 * Process a new chunk of data from an MPEG-TS stream. The chunks passed to this
+	 * function should be presented sequentially, and combine to form a coherent MPEG-TS
+	 * data stream, but they can be of arbitrary size.
+	 *
+	 * This method is called by a NodeJS Transform stream and so matches that spec
+	 *
+	 * @param  {Buffer} mpegtsData The latest sequential chunk of MPEG-TS data.
+	 * @param  {String} enc        The encoding of the passed data.
+	 * @param  {Function(err,result)} done     A `(err, result) => ...` callback.
 	 *
-	 * @param  {ReadableStream} readableStream The stream of data to be recorded.
-	 * @return {ReadableStream}                The input stream (for chaining)
 	 */
-	recordStream = (readableStream) => readableStream
-		.on('data', async (data) => {
-			readableStream.pause()
-			await this.ingestData(data)
-			readableStream.resume()
+	processMpegtsStreamData = (mpegtsData, enc, done) => {
+		this.mpegtsDemuxer.process(mpegtsData)
+		const demuxedPacket = this.getMostRecentDemuxedPacket() || generateEmptyPacket()
+		const payload = new Payload({
+			data: mpegtsData,
+			type: dataTypes.STREAM.CONTAINER,
+			duration: 0,
+			position: tsToMilliseconds(demuxedPacket.pts),
+			createdAt: Date.now(),
 		})
+		done(null, payload)
+	}
 
 	/**
 	 * Start the ingestion engine.
 	 *
-	 * @return {null}
+	 * This creates an FFmpeg process configured to convert input into a predetermined
+	 * format as specified by getFfmpegSettings, and announce new data to the ingestion
+	 * engine.
+	 *
+	 * It then spawns the input stream as specified by the implementing class, and creates
+	 * a pipeline that passes input->FFmpeg->packaging->ingestion
+	 *
+	 * @return {Stream} the ingestion pipeline that just got started
 	 */
 	start = () => {
-		const ffmpegSettings = this.getFfmpegSettings()
-		const ffmpegProcess = spawn('ffmpeg', ffmpegSettings)
+		this.ffmpegProcess = spawn('ffmpeg', this.getFfmpegSettings())
 		const inputStream = this.getInputStream()
-		if (inputStream !== null) {
-			inputStream.pipe(ffmpegProcess.stdin)
-		}
-		this.recordStream(ffmpegProcess.stdout)
+
 		logger.info(`Starting ingestion from ${this.constructor.name}...`)
+		inputStream.pipe(this.ffmpegProcess.stdin)
+		this.activeIngestionPipeline = pipeline(
+			this.ffmpegProcess.stdout,
+			this.mpegtsProcessingStream,
+			this.payloadIngestionStream,
+			() => this.stop(),
+		)
+		return this.activeIngestionPipeline
 	}
 
 	/**
-	 * Returns an ffmpeg settings array for this ingestion engine.
+	 * Stop the ingestion engine.
 	 *
-	 * @return {String[]} A list of ffmpeg command line parameters.
+	 * This stops the ingestion stream if it exists and terminates the FFmpeg process if it exists
 	 */
-	getFfmpegSettings = () => [
-		'-i', this.getInputLocation(),
-		...this.getTranscodeSettings(),
-		'-f', 'mpegts',
-		'-',
-	]
+	stop = () => {
+		if (this.activeIngestionPipeline !== null) {
+			this.activeIngestionPipeline.end()
```

Comment:
I'm not sure how to test this, but just to sanity-check: this doesn't create a loop, does it? My worry is that by calling `.end()` we may end up triggering this very `stop()` again, and unless something sets `this.activeIngestionPipeline = null` it will keep running. I read a bit about the streams API and couldn't be sure. If you're satisfied it won't (and/or that the tests prove it doesn't), I'll trust you!

Reply:
It's a good thought...

But this definitely won't create a callback loop.... because end is not actually a pipeline method, so it will error.  The test didn't catch this because we manually mocked a pipeline and included an `end` method.

I'm going to change this to trigger end on the input stream, which is the correct approach.  I did check and calling `.end()` twice on an input stream does not trigger the pipeline close callback more than once.

Reply:
Wow so I stumbled into helping by shining a light on it for a totally different reason! Like in a heist movie where the security guard drops his flashlight or soda can or something and it rolls behind the very spot the thieves are hiding and he discovers them while retrieving it and then they have to tie him up and escalate the stakes! Like that!!